File: MIT_18_335_introduction_to_numerical_methods/conjgrad.m
function xs=conjgrad(A,b,x0)
%CONJGRAD Conjugate Gradient (educational)

r0=b-A*x0;
x=x0;
r=r0;

ps=[];
xs=x;
i=0;
while norm(r)>norm(r0)*1e-3
  i=i+1;

  p0=r;
  p=p0;

  if i>1
    pk=rold;
    beta=-(p0'*A*pk)/(pk'*A*pk);
    p=p+beta*pk;
  end
  
  Ap=A*p;
  alpha=(p'*r)/(p'*Ap);
  x=x+alpha*p;
  rold=r;
  r=r-alpha*Ap;
 
  xs=[xs,x];
  ps=[ps,p];
end

File: Numerical_Eng_Python/conjGrad.py
## module conjGrad
''' x, numIter = conjGrad(Av,x,b,tol=1.0e-9)
    Conjugate gradient method for solving [A]{x} = {b}.
    The matrix [A] should be sparse. User must supply
    the function Av(v) that returns the vector [A]{v}.
'''    
from numarray import dot
from math import sqrt

def conjGrad(Av,x,b,tol=1.0e-9):
    n = len(b)
    r = b - Av(x)
    s = r.copy()
    for i in range(n):
        u = Av(s)
        alpha = dot(s,r)/dot(s,u)
        x = x + alpha*s
        r = b - Av(x)
        if(sqrt(dot(r,r))) < tol:
            break
        else:
            beta = -dot(r,u)/dot(s,u)
            s = r + beta*s
    return x,i


    
